pubsub.go

```diff
@@ -684,6 +758,50 @@ func (p *PubSub) Subscribe(topic string, opts ...SubOpt) (*Subscription, error)
 	return p.SubscribeByTopicDescriptor(&td, opts...)
 }
 
+type RelayOpt func(r *Relay) error
+
+// Join returns a new Relay for the given topic which will enable us to
+// propagate messages we receive to other peers without having to 'consume'/'subscribe' them.
+// For now, we ONLY support Relay without subscription.
+// We do NOT allow subscription without Relay.
+// Note that Join is not an instanteneous operation. It may take some time
+// before it is processed by the pubsub main loop and propagated to our peers.
+func (p *PubSub) Join(topic string, opts ...RelayOpt) (*Relay, error) {
```

Comment:
We need to create a `Relay` object every time 'Join' is called in order to reference count subscribers so we can announce we are leaving the topic if we have no more subscribers & un-cancelled Relay Only objects pending.  

However,

Reply:
We should have a single relay object and reference count; Subscribing increases the reference count, unsubscribe decreases.
In fact, I am not even sure we _need_ a relay object at all, just a reference count in the topic map.

Reply:
> In fact, I am not even sure we need a relay object at all, just a reference count in the topic map.

I don't even think we need a reference count at all. If we take the strategy from #198 and https://github.com/libp2p/go-libp2p-pubsub/pull/184#discussion_r330393936 then there should really only be one topic handle and that topic handle should be able to optionally be a relay.

Reply:
Yeah, that's a valid strategy.